### ISSUE TITLE
nice interface for a more low-level work

### ISSUE THREAD

**Issue by kilohsakul** [VeriFIT/mata#209]
E.g. in the new trim and variants revert. Things like push_back to ordvectors, filtering and renaming of moves and posts. What would be a nice interface for these things? The current one? What about e.g. non-const iterators? ...

**Comment by martinhruska** [comment]
Hmmm, if we want to replacing `operator[]` in the neighboring issue by something more explicit I would rather stay away from designing some some smart low level iterator here and use some set of the explicitly named functions. I think we should stay consistent in trade off between being explicit (and add a little bit bureaucracy) or being concise (assuming some knowledge of C++ constructs and idioms).

Anyway, a proper way to do this would be propably to wrap such disgraced structures by low level operations to some kind of monad and eventually unwrapping it later by making it consistent with the invariants (such as orderedness) again.

**Comment by Adda0** [comment]
> Anyway, a proper way to do this would be propably to wrap such disgraced structures by low level operations to some kind of monad and eventually unwrapping it later by making it consistent with the invariants (such as orderedness) again.

That is one possible approach. Depends on how we want to use these low-level functions later on. Do we want to be able to create algorithms using these functions individually? If so, we have to add these functions to the public interface as well.

**Comment by martinhruska** [comment]
yes, but the functions can return something like `struct RapedNfa` which would be defined like `struct RapedNfa { structu Nfa; T type_of_rape}`.

**Comment by Adda0** [comment]
That could be. Depends on how we intend to use these functions. Only upon considering the use cases can we decide on a reasonable approach.
